File: bookstore-go/internal/handler/review_test.go
package handler_test

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestListReviews(t *testing.T) {
	mux, books, authors, reviews := newTestMux(t)

	a, _ := authors.Create("Author", "")
	b, _ := books.Create("Book", a.ID, "", 2000)
	reviews.Create(b.ID, 5, "Great read.")
	reviews.Create(b.ID, 3, "It was okay.")

	req := httptest.NewRequest(http.MethodGet, fmt.Sprintf("/api/books/%d/reviews", b.ID), nil)
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusOK {
		t.Errorf("status = %d, want 200", w.Code)
	}
	var result []any
	if err := json.NewDecoder(w.Body).Decode(&result); err != nil {
		t.Fatalf("decode: %v", err)
	}
	if len(result) != 2 {
		t.Errorf("len = %d, want 2", len(result))
	}
}

func TestCreateReview(t *testing.T) {
	mux, books, authors, _ := newTestMux(t)

	a, _ := authors.Create("Author", "")
	b, _ := books.Create("Book", a.ID, "", 2000)

	body, _ := json.Marshal(map[string]any{"rating": 4, "review_text": "Solid book."})
	req := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/api/books/%d/reviews", b.ID), bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusCreated {
		t.Errorf("status = %d, want 201", w.Code)
	}
}

// TestCreateReviewNonexistentBook: posting a review for a non-existent book should return 404.
func TestCreateReviewNonexistentBook(t *testing.T) {
	mux, _, _, _ := newTestMux(t)

	body, _ := json.Marshal(map[string]any{"rating": 5, "review_text": "Orphaned review."})
	req := httptest.NewRequest(http.MethodPost, "/api/books/99999/reviews", bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusNotFound {
		t.Errorf("expected 404 for review on non-existent book, got %d", w.Code)
	}
}

// TestCreateReviewValidation: invalid rating and review_text values should return 400.
func TestCreateReviewValidation(t *testing.T) {
	mux, books, authors, _ := newTestMux(t)

	a, _ := authors.Create("Author", "")
	b, _ := books.Create("Book", a.ID, "", 2000)

	cases := []struct {
		name   string
		rating int
		text   string
	}{
		{"zero rating", 0, "Some text here."},
		{"negative rating", -1, "Some text here."},
		{"huge rating", 999, "Some text here."},
		{"empty text", 3, ""},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			body, _ := json.Marshal(map[string]any{"rating": tc.rating, "review_text": tc.text})
			req := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/api/books/%d/reviews", b.ID), bytes.NewReader(body))
			req.Header.Set("Content-Type", "application/json")
			w := httptest.NewRecorder()
			mux.ServeHTTP(w, req)

			if w.Code != http.StatusBadRequest {
				t.Errorf("expected 400 for invalid input (%s), got %d", tc.name, w.Code)
			}
		})
	}
}

File: bookstore-ts/tests/handler/review.test.ts
import { expect, test, describe, afterEach } from "bun:test";
import { newTestServer } from "../helpers";

let env: ReturnType<typeof newTestServer> | null = null;

afterEach(() => {
  env?.stop();
  env = null;
});

function setup() {
  env = newTestServer();
  return env;
}

describe("ListReviews", () => {
  test("returns 200 and reviews", async () => {
    const { base, books, authors, reviews } = setup();
    const a = authors.create("Author", "");
    const b = books.create("Book", a.id, "", 2000);
    reviews.create(b.id, 5, "Great read.");
    reviews.create(b.id, 3, "It was okay.");
    const res = await fetch(`${base}/api/books/${b.id}/reviews`);
    expect(res.status).toBe(200);
    const body = (await res.json()) as unknown[];
    expect(body.length).toBe(2);
  });
});

test("CreateReview returns 201", async () => {
  const { base, books, authors } = setup();
  const a = authors.create("Author", "");
  const b = books.create("Book", a.id, "", 2000);
  const res = await fetch(`${base}/api/books/${b.id}/reviews`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ rating: 4, review_text: "Solid book." }),
  });
  expect(res.status).toBe(201);
});

// posting a review for a non-existent book should return 404.
test("CreateReview for non-existent book returns 404", async () => {
  const { base } = setup();
  const res = await fetch(`${base}/api/books/99999/reviews`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ rating: 5, review_text: "Orphaned review." }),
  });
  expect(res.status).toBe(404);
});

// invalid rating and review_text values should return 400.
describe("CreateReview validation", () => {
  const cases: { name: string; rating: number; text: string }[] = [
    { name: "zero rating", rating: 0, text: "Some text here." },
    { name: "negative rating", rating: -1, text: "Some text here." },
    { name: "huge rating", rating: 999, text: "Some text here." },
    { name: "empty text", rating: 3, text: "" },
  ];

  for (const tc of cases) {
    test(tc.name, async () => {
      const { base, books, authors } = setup();
      const a = authors.create("Author", "");
      const b = books.create("Book", a.id, "", 2000);
      const res = await fetch(`${base}/api/books/${b.id}/reviews`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ rating: tc.rating, review_text: tc.text }),
      });
      expect(res.status).toBe(400);
    });
  }
});
